internal/pkg/service/stream/sink/router/router.go

```diff
@@ -0,0 +1,338 @@
+// Router routes the record to the desired sink pipeline.
+type Router interface {
+	DispatchToSources(sources []key.SourceKey, c recordctx.Context) SourcesResult
+	DispatchToSource(sourceKey key.SourceKey, c recordctx.Context) SourceResult
+	DispatchToSink(sinkKey key.SinkKey, c recordctx.Context) SinkResult
+}
+
+type router struct {
+	logger      log.Logger
+	plugins     *plugin.Plugins
+	definitions *definitionRepo.Repository
+	// sinks field contains in-memory snapshot of all active sinks. Only necessary data is saved.
+	sinks *etcdop.Mirror[definition.Sink, *sinkData]
+	// closed channel block new writer during shutdown
+	closed chan struct{}
+	// wg waits for all goroutines
+	wg sync.WaitGroup
+	// lock protects the pipelines map.
+	lock sync.Mutex
+	// pipelines - map of active pipelines per sink
+	pipelines map[key.SinkKey]*pipelineRef
+}
+
+type sinkData struct {
+	sinkKey key.SinkKey
+	enabled bool
+}
+
+type dependencies interface {
+	Logger() log.Logger
+	Process() *servicectx.Process
+	Plugins() *plugin.Plugins
+	DefinitionRepository() *definitionRepo.Repository
+}
+
+func New(d dependencies) (Router, error) {
+	r := &router{
+		logger:      d.Logger().WithComponent("sink.router"),
+		plugins:     d.Plugins(),
+		definitions: d.DefinitionRepository(),
+		closed:      make(chan struct{}),
+		pipelines:   make(map[key.SinkKey]*pipelineRef),
+	}
+
+	ctx, cancelMirror := context.WithCancel(context.Background())
+	d.Process().OnShutdown(func(ctx context.Context) {
+		r.logger.Infof(ctx, "shutting down sink router")
+
+		// Block new writes
+		close(r.closed)
+
+		// Stop mirroring
+		cancelMirror()
+
+		// Wait for in-flight writes
+		r.wg.Wait()
+
+		// Wait for closing all pipelines
+		r.closeAllPipelines(ctx, "shutdown")
+		r.wg.Wait()
+
+		r.logger.Infof(ctx, "sink router shutdown done")
+	})
+
+	// Start sinks mirroring, only necessary data is saved
+	{
+		var errCh <-chan error
+		r.sinks, errCh = etcdop.
+			SetupMirror(
+				r.logger,
+				r.definitions.Sink().GetAllAndWatch(ctx, clientv3.WithPrevKV()),
+				func(kv *op.KeyValue, sink definition.Sink) string {
+					return sink.SinkKey.String()
+				},
+				func(kv *op.KeyValue, sink definition.Sink) *sinkData {
+					return &sinkData{
+						sinkKey: sink.SinkKey,
+						enabled: sink.IsEnabled(),
+					}
+				},
+			).
+			WithOnUpdate(func(changes etcdop.MirrorUpdatedKeys[*sinkData]) {
+				// Close updated sinks, the pipeline must be re-created.
+				// Closing the old pipeline blocks the creation of a new one.
+				for _, kv := range changes.Updated {
+					r.closePipeline(ctx, kv.Value.sinkKey, "sink updated")
+				}
+				// Closed delete sinks
+				for _, kv := range changes.Deleted {
+					r.closePipeline(ctx, kv.Value.sinkKey, "sink deleted")
+				}
+			}).
+			StartMirroring(ctx, &r.wg)
+		if err := <-errCh; err != nil {
+			return nil, err
+		}
+	}
+
+	return r, nil
+}
+
+func (r *router) DispatchToSources(sources []key.SourceKey, c recordctx.Context) SourcesResult {
```

Comment:
why do we have DispatchToSources and DispatchToSource? When do we use Sources and when Source?

Reply:
We now , use the top level `DispatchToSources` method now, but in the future, and for tests, also other make sense.

---------

HTTP source:

- `SourceKey` - contains branch ID
  - It can have different settings per branch.
- But, HTTP source endpoint is one, for all branches: `/stream/<projectID/<sourceID>/<secret>`.
- And the record should be dispatched to all branches (so to all `Sources` entities), to all sinks in the sources.

So there are 3 layers:
- `Sources` (branches)
- `Souces` (per branch)
- `Sink` (per branch and source)